homeassistant/components/lametric/config_flow.py

```diff
@@ -0,0 +1,279 @@
+    SelectOptionDict,
+    SelectSelector,
+    SelectSelectorConfig,
+    SelectSelectorMode,
+    TextSelector,
+    TextSelectorConfig,
+    TextSelectorType,
+)
+from homeassistant.util.network import is_link_local
+
+from .const import DOMAIN, LOGGER
+
+
+class LaMetricFlowHandler(AbstractOAuth2FlowHandler, domain=DOMAIN):
+    """Handle a LaMetric config flow."""
+
+    DOMAIN = DOMAIN
+    VERSION = 1
+
+    devices: dict[str, CloudDevice]
+    discovered_host: str
+    discovered_serial: str
+    discovered: bool = False
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return LOGGER
+
+    @property
+    def extra_authorize_data(self) -> dict[str, Any]:
+        """Extra data that needs to be appended to the authorize url."""
+        return {"scope": "basic devices_read"}
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initiated by the user."""
+        return await self.async_step_choice_enter_manual_or_fetch_cloud()
+
+    async def async_step_ssdp(self, discovery_info: SsdpServiceInfo) -> FlowResult:
+        """Handle a flow initiated by SSDP discovery."""
+        url = URL(discovery_info.ssdp_location or "")
+        if url.host is None or not (
+            serial := discovery_info.upnp.get(ATTR_UPNP_SERIAL)
+        ):
+            return self.async_abort(reason="invalid_discovery_info")
+
+        if is_link_local(ip_address(url.host)):
+            return self.async_abort(reason="link_local_address")
+
+        await self.async_set_unique_id(serial)
+        self._abort_if_unique_id_configured(updates={CONF_HOST: url.host})
+
+        self.context.update(
+            {
+                "title_placeholders": {
+                    "name": discovery_info.upnp.get(
+                        ATTR_UPNP_FRIENDLY_NAME, "LaMetric TIME"
+                    ),
+                },
+                "configuration_url": "https://developer.lametric.com",
+            }
+        )
+
+        self.discovered = True
+        self.discovered_host = str(url.host)
+        self.discovered_serial = serial
+        return await self.async_step_choice_enter_manual_or_fetch_cloud()
+
+    async def async_step_choice_enter_manual_or_fetch_cloud(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the user's choice of entering the manual credentials or fetching the cloud credentials."""
+        if user_input is not None:
+
+            if user_input[CONF_METHOD] == "cloud":
+                return await self.async_step_pick_implementation()
+            return await self.async_step_manual_entry()
+
+        return self.async_show_form(
+            step_id="choice_enter_manual_or_fetch_cloud",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_METHOD, default="cloud"): SelectSelector(
```

Comment:
Maybe use a `async_show_menu` instead?

Reply:
Oh forgot about that one! That is a good idea ❤️